openedx/core/djangoapps/content_libraries/tasks.py

```diff
@@ -39,10 +62,197 @@
+        # This will be true if the entity was [soft] deleted, but we're now reverting that deletion:
+        is_undeleted = (record.old_version is None and record.new_version is not None)
+        # This will be true if the entity was created and we're now deleting it by reverting that creation:
+        is_deleted = (record.old_version is not None and record.new_version is None)
+        if hasattr(record.entity, "component"):
+            usage_key = api.library_component_usage_key(library_key, record.entity.component)
+            event = LIBRARY_BLOCK_UPDATED
+            if is_deleted:
+                event = LIBRARY_BLOCK_DELETED
+            elif is_undeleted:
+                event = LIBRARY_BLOCK_CREATED
+            event.send_event(library_block=LibraryBlockData(library_key=library_key, usage_key=usage_key))
+            # If any containers contain this component, their child list / component count may need to be updated
+            # e.g. if this was a newly created component in the container and is now deleted, or this was deleted and
+            # is now restored.
+            for parent_container in api.get_containers_contains_component(usage_key):
+                updated_container_keys.add(parent_container.container_key)
+
+            # TODO: do we also need to send CONTENT_OBJECT_ASSOCIATIONS_CHANGED for this component, or is
+            # LIBRARY_BLOCK_UPDATED sufficient?
+        elif hasattr(record.entity, "container"):
+            container_key = api.library_container_locator(library_key, record.entity.container)
+            if is_deleted:
+                deleted_container_keys.add(container_key)
+            elif is_undeleted:
+                created_container_keys.add(container_key)
+            else:
+                updated_container_keys.add(container_key)
+        else:
+            log.warning(
+                f"PublishableEntity {record.entity.pk} / {record.entity.key} was modified during publish operation "
+                "but is of unknown type."
+            )
+        # If any collections contain this entity, their item count may need to be updated, e.g. if this was a
+        # newly created component in the collection and is now deleted, or this was deleted and is now re-added.
+        for parent_collection in authoring_api.get_entity_collections(
+            record.entity.learning_package_id, record.entity.key,
+        ):
+            collection_key = api.library_collection_locator(
+                library_key=library_key,
+                collection_key=parent_collection.key,
+            )
+            affected_collection_keys.add(collection_key)
+
+    for container_key in deleted_container_keys:
+        LIBRARY_CONTAINER_DELETED.send_event(
+            library_container=LibraryContainerData(container_key=container_key)
+        )
+        # Don't bother sending UPDATED events for these containers that are now deleted
+        created_container_keys.remove(container_key)
```

Comment:
`remove` raises error if key is not in the set. 

You can verify this by creating a new container (do not publish) and creating a random component (**required for discard option to be enabled, for some reason changes to unit does not enable publish or discard button**) and using discard option.


```suggestion
        created_container_keys.discard(container_key)
```

Reply:
Ooh, thanks for catching that!